### PR TITLE
Add support for unread thread notifications

### PR DESCRIPTION
**Pull request by gsouquet**
Implements https://github.com/matrix-org/matrix-spec-proposals/pull/3773
Related to https://github.com/vector-im/element-web/issues/23192

Passes the quality gate but still has optional strict null checks failures

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add support for unread thread notifications ([\#2726](https://github.com/matrix-org/matrix-js-sdk/pull/2726)).

### CODE REVIEW

File: src/sync.ts [hunk 1]
```diff
@@ -1264,6 +1268,28 @@ export class SyncApi {
                 }
             }
 
+            room.resetThreadUnreadNotificationCount();
+            const unreadThreadNotifications = joinObj[UNREAD_THREAD_NOTIFICATIONS.name];
```
Comment (weeman1337) [inline src/sync.ts:1272]: Shouldn't we try to retrieve the value by `altName` as well?

File: src/sync.ts [hunk 2]
```diff
@@ -1264,6 +1268,28 @@ export class SyncApi {
...
+            if (unreadThreadNotifications) {
+                Object.entries(unreadThreadNotifications).forEach(([threadId, unreadNotification]) => {
+                    room.setThreadUnreadNotificationCount(
+                        threadId,
+                        NotificationCountType.Total,
+                        unreadNotification.notification_count,
+                    );
+
+                    const hasUnreadNotification =
+                        room.getThreadUnreadNotificationCount(threadId, NotificationCountType.Highlight) <= 0;
```
Comment (weeman1337) [inline src/sync.ts:1282]: Is `<= 0` correct here? I would expect to check for `> 0`.

**Review by weeman1337 — COMMENTED** [review]
:+1: just some small questions

File: src/sync.ts (hunk 2 above)
Comment (gsouquet) [reply src/sync.ts:1282]: I was confused when I named that variable. I replicated the logic from above
https://github.com/matrix-org/matrix-js-sdk/blob/7d5adf48bcca8fcdcefea707528aa906c71929e2/src/sync.ts#L1263

**Review by weeman1337 — APPROVED** [review]

**Comment by t3chguy** [comment]
This breaks initialSync filtering support

https://github.com/matrix-org/matrix-js-sdk/commit/21a6f61b7b7bccb899c2aa34e6c881b729524bcd#diff-1b0cb46cd1fcdaacf73c771bf7eed6e00a12ee030da1dbcca4ae12842e1ecd81R709-R712

image

This may need reverting if not remedied

**Comment by t3chguy** [comment]
Inherently this is broken

1. it never sets the unstable-prefixed filter
2. it sets the filter with a value of `false` which is redundant
3. it only sets it for the initial sync and not any later syncs which means the counts won't ever update on incremental/gappy syncs.

**Pull request merged**